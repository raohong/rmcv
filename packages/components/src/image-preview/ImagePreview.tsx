import { SpringConfig, useSpring } from '@react-spring/web';
import {
  useControllableValue,
  useEventCallback,
  useLockScroll,
  useMeasure,
  useMergeRefs,
  useUnmountedRef,
  useUpdateEffect,
} from '@rmc-vant/hooks';
import { noop } from '@rmc-vant/utils';
import { rubberbandIfOutOfBounds } from '@use-gesture/react';
import React, { useEffect, useMemo, useRef, useState } from 'react';
import { getDataOrAriaProps } from '../_utils';
import { useThemeProps } from '../config-provider';
import Portal from '../portal';
import ImagePreviewItem from './ImagePreviewItem';
import { ImagePreviewName, composeImagePreviewSlotClassNames } from './classNames';
import type {
  DragExitHandler,
  ImagePreviewComponentState,
  ImagePreviewProps,
} from './interface';
import {
  ImagePreviewContent,
  ImagePreviewHeader,
  ImagePreviewIndex,
  ImagePreviewList,
  ImagePreviewOverlay,
  ImagePreviewRoot,
} from './styles';
import { clampTo } from './utils';

const VisibleAnimationSpringConfig: SpringConfig = {
  tension: 500,
  friction: 40,
  precision: 0.1,
};
const XOffsetAnimationSpringConfig: SpringConfig = {
  tension: 450,
  friction: 40,
  precision: 0.1,
};

const ImagePreview: React.FC<ImagePreviewProps> = (props) => {
  const {
    onClose,
    teleport,
    classNames,
    open = false,
    showIndex = true,
    lazyRender = true,
    afterClose = noop,
    maxZoom = 3,
    minZoom = 1 / 3,
    images = [],
    ...rest
  } = useThemeProps(ImagePreviewName, props);
  const [activeIndex, setActiveIndex] = useControllableValue(rest, {
    valuePropName: 'activeIndex',
    defaultValuePropName: 'defaultActiveIndex',
    defaultValue: 0,
    format: (val) => clampTo(0, val ?? 0, images.length - 1),
  });
  const [animateActiveIndex, setAnimateActiveIndex] = useState(activeIndex);
  const unmountedRef = useUnmountedRef();
  const [isAnimate, setIsAnimate] = useState(true);
  const lockRef = useLockScroll(open);
  const {
    setRef,
    data: { width, height },
  } = useMeasure();
  const mergedContentRef = useMergeRefs(lockRef, setRef);

  const [animateVisible, setAnimateVisible] = useState(!!open);
  const [{ x }, xCtrl] = useSpring(() => ({
    x: 0,
    config: {
      precision: VisibleAnimationSpringConfig.precision,
    },
  }));
  const [{ y, progress }, ctrl] = useSpring(() => ({
    y: 0,
    progress: 0,
    config: VisibleAnimationSpringConfig,
  }));
  const lastVelocity = useRef(0);
  const afterCloseCallback = useEventCallback(afterClose);

  const componentState = useMemo<ImagePreviewComponentState>(() => ({}), []);
  const slotClassNames = composeImagePreviewSlotClassNames(
    componentState,
    classNames,
  );

  const handleTap = () => {
    onClose?.();
  };

  const handleDragExit: DragExitHandler = ({
    movement,
    last,
    velocity,
    direction,
  }) => {
    const dragProgress = movement >= 0 ? Math.min(1, movement / height) : 0;

    if (!last) {
      ctrl.set({ progress: 1 - dragProgress, y: movement });

      return;
    }

    if ((movement >= height / 6 || Math.abs(velocity) >= 1) && direction >= 0) {
      onClose?.();
    } else {
      ctrl.start({
        progress: 1,
        y: 0,
        config: {
          velocity,
        },
      });
    }
  };

  const handleDrag = (
    movement: number,
    dir: number,
    velocity: number,
    last: boolean,
  ) => {
    const MOVEMENT_BOUNDARY = width / 3;
    const VELOCITY_BOUNDARY = 0.7;
    const len = images.length;

    let val = -width * activeIndex + movement;
    val = rubberbandIfOutOfBounds(val, -width * (len - 1), 0);

    if (!last) {
      xCtrl.set({ x: val });
      return;
    }

    let nextActiveIndex = activeIndex;
    let v = 0;

    if (Math.abs(movement) >= MOVEMENT_BOUNDARY) {
      nextActiveIndex += -Math.sign(dir);
    } else if (Math.abs(velocity) >= VELOCITY_BOUNDARY) {
      nextActiveIndex += -Math.sign(dir);
      v = velocity;
    }

    nextActiveIndex = clampTo(nextActiveIndex, 0, len - 1);

    if (activeIndex !== nextActiveIndex) {
      setActiveIndex(nextActiveIndex);
      lastVelocity.current = v;
    } else {
      xCtrl.start({ x: nextActiveIndex * -width });
    }
  };

  useUpdateEffect(() => {
    if (open) {
      setAnimateVisible(true);
    }
  }, [open]);

  useEffect(() => {
    if (!width || activeIndex === animateActiveIndex) {
      return;
    }

    const dest = activeIndex * -width;

    if (dest === x.get()) {
      return;
    }

    // visible 由 false => true 时 x 关闭动画
    if (!isAnimate) {
      xCtrl.set({
        x: dest,
      });
      setAnimateActiveIndex(activeIndex);
      setIsAnimate(true);
    } else {
      xCtrl.start({
        x: dest,
        config: {
          ...XOffsetAnimationSpringConfig,
          velocity: lastVelocity.current,
        },
        onRest({ finished }) {
          if (finished && !unmountedRef.current) {
            setAnimateActiveIndex(activeIndex);
          }
        },
      });

      lastVelocity.current = 0;
    }
  }, [activeIndex, width, xCtrl, x, isAnimate, animateActiveIndex, unmountedRef]);

  useEffect(() => {
    if (height === 0) {
      return;
    }

    ctrl.start({
      y: open ? 0 : height,
      progress: open ? 1 : 0,
      onRest: {
        y({ finished }) {
          if (!unmountedRef.current && finished && !open) {
            setAnimateVisible(false);
          }
        },
      },
    });
  }, [open, ctrl, height, unmountedRef]);

  useUpdateEffect(() => {
    if (!animateVisible) {
      afterCloseCallback();
      setIsAnimate(false);
    }
  }, [afterCloseCallback, animateVisible]);

  const content = (
    <Portal disablePortal={!teleport} teleport={teleport}>
      <ImagePreviewRoot
        className={slotClassNames.root}
        componentState={componentState}
        {...getDataOrAriaProps(rest)}
      >
        <ImagePreviewOverlay
          style={{
            opacity: progress,
          }}
        />
        <ImagePreviewContent ref={mergedContentRef}>
          {!!showIndex && (
            <ImagePreviewHeader
              style={{
                y: progress.to([0, 1], ['-100%', '0%'], 'clamp'),
              }}
              className={slotClassNames.header}
              componentState={componentState}
            >
              <ImagePreviewIndex
                componentState={componentState}
                className={slotClassNames.index}
              >
                {activeIndex + 1} / {images.length}
              </ImagePreviewIndex>
            </ImagePreviewHeader>
          )}
          <ImagePreviewList
            style={{
              width: `${100 * images.length}vw`,
              x,
              y,
            }}
          >
            {images.map((item, index) => (
              <ImagePreviewItem
                // eslint-disable-next-line react/no-array-index-key
                key={index}
                imageUrl={item}
                containerHeight={height}
                containerWidth={width}
                onDrag={handleDrag}
                onDragExit={handleDragExit}
                onTap={handleTap}
                maxScale={maxZoom}
                minScale={minZoom}
                visible={animateActiveIndex === index}
                gestureEnabled={activeIndex === animateActiveIndex}
                slotClassNames={slotClassNames}
                componentState={componentState}
              />
            ))}
          </ImagePreviewList>
        </ImagePreviewContent>
      </ImagePreviewRoot>
    </Portal>
  );

  return <>{lazyRender ? (animateVisible ? content : null) : content}</>;
};

export default ImagePreview;
