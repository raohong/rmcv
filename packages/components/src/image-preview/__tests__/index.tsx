import { render } from '@testing-library/react';
import React from 'react';
import ImagePreview, { ImagePreviewComponent } from '../';

const list = [
  'https://cdn.jsdelivr.net/npm/@vant/assets/apple-1.jpeg',
  'https://cdn.jsdelivr.net/npm/@vant/assets/apple-2.jpeg',
  'https://cdn.jsdelivr.net/npm/@vant/assets/apple-3.jpeg',
  'https://cdn.jsdelivr.net/npm/@vant/assets/apple-4.jpeg',
];

it('snapshot', () => {
  const tree = render(
    <ImagePreviewComponent images={list} lazyRender={false} visible />,
  );
  expect(tree.asFragment()).toMatchSnapshot();
});

it('impertive call', () => {
  const instance = ImagePreview(list);

  expect(instance.current?.swipeTo).toBeDefined();
});
