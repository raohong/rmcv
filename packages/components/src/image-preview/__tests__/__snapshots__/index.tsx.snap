// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-font-smooth: antialiased;
  position: relative;
  touch-action: none;
  list-style: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,Helvetica Neue,Helvetica,Segoe UI,Arial,Roboto,PingFang SC,miui,Hiragino Sans GB,Microsoft Yahei,sans-serif;
}

.emotion-1 {
  position: fixed;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  will-change: opacity;
  background: rgba(0, 0, 0, 1);
  z-index: 1;
}

.emotion-2 {
  position: fixed;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  z-index: 1;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 1;
  height: 52px;
}

.emotion-4 {
  color: #fff;
  font-size: 14px;
  line-height: 20px;
  text-shadow: 0 1px 1px #323233;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: hidden;
  height: 100%;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100vw;
  height: 100%;
  touch-action: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-7 {
  transform-origin: center;
  will-change: transform;
  width: 100%;
}

.emotion-8 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-font-smooth: antialiased;
  position: relative;
  touch-action: none;
  list-style: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,Helvetica Neue,Helvetica,Segoe UI,Arial,Roboto,PingFang SC,miui,Hiragino Sans GB,Microsoft Yahei,sans-serif;
  display: inline-block;
  min-width: 1px;
  min-height: 1px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: block;
}

.emotion-8 img {
  height: auto;
}

.emotion-9 {
  box-sizing: border-box;
  display: block;
  width: 100%;
  height: 100%;
}

<div
    class="RVuiImagePreview-root emotion-0"
  >
    <div
      class="emotion-1"
      style="opacity: 0;"
    />
    <div
      class="emotion-2"
    >
      <div
        class="RVuiImagePreview-header emotion-3"
        style="transform: translate3d(0,-100%,0);"
      >
        <span
          class="RVuiImagePreview-index emotion-4"
        >
          1 /4
        </span>
      </div>
      <div
        class="emotion-5"
        style="width: 400vw; transform: none;"
      >
        <div
          class="RVuiImagePreview-item emotion-6"
        >
          <div
            class="emotion-7"
            style="transform: translate3d(0px, 0px, 0) scale(1);"
          >
            <div
              class="RVuiImage-root RVuiImage-statusNone RVuiImagePreview-image emotion-8"
              style="width: 100%;"
            >
              <img
                alt=""
                class="RVuiImage-img emotion-9"
                draggable="false"
                src="https://cdn.jsdelivr.net/npm/@vant/assets/apple-1.jpeg"
              />
            </div>
          </div>
        </div>
        <div
          class="RVuiImagePreview-item emotion-6"
        >
          <div
            class="emotion-7"
            style="transform: translate3d(0px, 0px, 0) scale(1);"
          >
            <div
              class="RVuiImage-root RVuiImage-statusNone RVuiImagePreview-image emotion-8"
              style="width: 100%;"
            >
              <img
                alt=""
                class="RVuiImage-img emotion-9"
                draggable="false"
                src="https://cdn.jsdelivr.net/npm/@vant/assets/apple-2.jpeg"
              />
            </div>
          </div>
        </div>
        <div
          class="RVuiImagePreview-item emotion-6"
        >
          <div
            class="emotion-7"
            style="transform: translate3d(0px, 0px, 0) scale(1);"
          >
            <div
              class="RVuiImage-root RVuiImage-statusNone RVuiImagePreview-image emotion-8"
              style="width: 100%;"
            >
              <img
                alt=""
                class="RVuiImage-img emotion-9"
                draggable="false"
                src="https://cdn.jsdelivr.net/npm/@vant/assets/apple-3.jpeg"
              />
            </div>
          </div>
        </div>
        <div
          class="RVuiImagePreview-item emotion-6"
        >
          <div
            class="emotion-7"
            style="transform: translate3d(0px, 0px, 0) scale(1);"
          >
            <div
              class="RVuiImage-root RVuiImage-statusNone RVuiImagePreview-image emotion-8"
              style="width: 100%;"
            >
              <img
                alt=""
                class="RVuiImage-img emotion-9"
                draggable="false"
                src="https://cdn.jsdelivr.net/npm/@vant/assets/apple-4.jpeg"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
